File: heony704/1_two-sum.js
/**
 * https://leetcode.com/problems/two-sum/
 * 9m
 *
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
// 배열을 순회하면서 숫자를 저장하고, 저장한 값 중에 정답(target - 현재숫자)이 있는지 확인
var twoSum = function (nums, target) {
  // Map에 숫자, 인덱스 저장
  const numWithIndex = new Map();

  for (let i = 0; i < nums.length; i++) {
    if (numWithIndex.has(target - nums[i])) {
      return [numWithIndex.get(target - nums[i]), i];
    }

    numWithIndex.set(nums[i], i);
  }

  return [];
};

var twoSum = function (nums, target) {
  // nums 배열 앞에서부터 일일이 더해가며 계산
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) return [i, j];
    }
  }
};

File: citrussoda/4week/cpp/1_two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> m;

        for (int i = 0; i < nums.size(); i++) {
            // if (m.find(target - nums[i]) != m.end()) {
            // count 가 읽기도 쉽고 더 빠르다!! <= 틀린내용 다만 큰 차이는 없음 알고리즘에 따라 다른 것 같다. 혹은 set과 map에 따라서도 다른듯.
            if (m.count(target - nums[i])) {
                return {m[target - nums[i]], i};
            }

            m[nums[i]] = i;
        }

        return {};
    }
};